src/main/scala/com/wavesplatform/state2/BlockchainUpdaterImpl.scala

```diff
@@ -76,112 +72,153 @@ class BlockchainUpdaterImpl private(persisted: StateWriter with StateReader,
       bottomMemoryDiff.set(topMemoryDiff())
       topMemoryDiff.set(BlockDiff.empty)
     }
-    (liquidBlockCandidatesDiff().get(block.reference) match {
-      case Some(referencedLiquidDiff) =>
-        val asFirmBlock = referencedLiquidDiff.copy(heightDiff = 1)
-        ngHistoryWriter.appendBlock(block)(BlockDiffer.fromBlock(settings,
-          composite(currentPersistedBlocksState, () => asFirmBlock),
-          ngHistoryWriter.forgeBlock(block.reference).map(_._1), block)).map { case ((newBlockDiff, discraded)) =>
-          topMemoryDiff.set(Monoid.combine(topMemoryDiff(), asFirmBlock))
-          liquidBlockCandidatesDiff.set(Map(block.uniqueId -> newBlockDiff))
-          discraded
-        }
+    (ngState() match {
       case None =>
-        ngHistoryWriter.appendBlock(block)(BlockDiffer.fromBlock(
-          settings, currentPersistedBlocksState, ngHistoryWriter.lastBlock, block)).map { case ((newBlockDiff, discraded)) =>
-          liquidBlockCandidatesDiff.set(Map(block.uniqueId -> newBlockDiff))
-          discraded
+        lazy val logDetails = s"The referenced block(${block.reference})" +
+          s" ${if (historyWriter.contains(block.reference)) "exits, it's not last persisted" else "doesn't exist"}"
+        historyWriter.lastBlock match {
+          case Some(lastInner) if lastInner.uniqueId != block.reference =>
+            Left(BlockAppendError(s"References incorrect or non-existing block: " + logDetails, block))
+          case _ =>
+            historyWriter.lastBlock.map(_.uniqueId) match {
```

Comment:
Looks strange, because we already in the None branch. So, we fall to
```scala
case _ => BlockDiffer.fromBlock(settings,
```
here. What do you think?

Reply:
We are in branch(ng=none, lastblock=none_or_matches_by_id), but you're right, will simplify